hibernate-core/src/main/java/org/hibernate/hql/spi/id/cte/CteValuesListBuilder.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.hql.spi.id.cte;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Builds the CTE with VALUES list clause that wraps the identifiers to be updated/deleted.
+ *
+ * @author Evandro Pires da Silva
+ * @author Vlad Mihalcea
+ */
+public class CteValuesListBuilder {
+
+	private final String tableName;
+
+	private final String[] columns;
+
+	private final List<Object[]> ids;
+
+	private StringBuilder statementBuilder;
+
+	public CteValuesListBuilder(
+			String tableName,
+			String[] columns,
+			List<Object[]> ids) {
+		this.tableName = tableName;
+		this.columns = columns;
+		this.ids = ids;
+
+		this.statementBuilder = buildStatement();
+	}
+
+	public String toStatement(String statement) {
+		return statementBuilder + statement;
```

Comment:
Actually, this is the only place where the `StringBuilder` makes sense since we can reuse it for multiple SQL statements. For the inline bulk-id strategies, your observation is very good, so I removed the buldStatement entirely and just build it in `toStatement` since the method is called only once.